### PR TITLE
[16.0][sale_order_currency_rate] New module that show currency rate on sale orders

### PR DESCRIPTION
**Pull request by oscarjarsa**

### CODE REVIEW

**Comment by oscarjarsa** [comment]
@alan196 I already added the two fields to sale currency rate and inverse_currency_rate but I don't know if what I did in code is correct

**Comment by rafaelbn** [comment]
Hello @oscarjarsa you are modiying here many modules

- sale_order_currency_rate
- sale_partner_pricelist

Please one PR per module and a little explanation of the motivation in the commit message

Test are 🔴 

Thank you!

File: sale_order_currency_rate/models/res_config_settings.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Copyright 2023 Jarsa
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl).
+
+from odoo import fields, models
+
+
+class ResConfigSettings(models.TransientModel):
+    _inherit = "res.config.settings"
+
+    sale_show_currency_rate = fields.Selection(
```
Comment (alan196) [inline sale_order_currency_rate/models/res_config_settings.py:10]:
```suggestion
    sale_show_currency_rate = fields.Selection(related="company_id.sale_show_currency_rate")
```

You don't need to add the selections and the string, it will inherit from the relation

File: sale_order_currency_rate/models/sale_order.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+from odoo import api, fields, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    currency_rate = fields.Float(
+        compute="_compute_currency_rate",
+        string="Currency",
+        readonly=True,
```
Comment (alan196) [inline sale_order_currency_rate/models/sale_order.py]:
```suggestion
```

If it's a computed field, the default behavior is readonly True

File: sale_order_currency_rate/models/sale_order.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        store=True,
+    )
+
+    inverse_currency_rate = fields.Float(
+        compute="_compute_currency_rate",
+        string="Inverse Currency",
+        readonly=True,
```
Comment (alan196) [inline sale_order_currency_rate/models/sale_order.py]:
```suggestion
```

Same

File: sale_order_currency_rate/readme/DESCRIPTION.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Description
```
Comment (alan196) [inline sale_order_currency_rate/readme/DESCRIPTION.rst]: Please add a description of what the module do.

File: sale_order_currency_rate/readme/USAGE.rst [hunk 5]
```diff
@@ -0,0 +1,2 @@
+1. Go to the Customers menu.
+2. Select the customer/partner for whom you want to set the allowed pricelist.
```
Comment (alan196) [inline sale_order_currency_rate/readme/USAGE.rst]: This usage is not related with this module

File: sale_order_currency_rate/views/sale_order.xml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="view_order_form" model="ir.ui.view">
+        <field name="name">sale.order.form</field>
+        <field name="model">sale.order</field>
+        <field name="inherit_id" ref="sale.view_order_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='currency_id']" position="after">
+                <field name="currency_rate" />
+                <field name="inverse_currency_rate" />
```
Comment (alan196) [inline sale_order_currency_rate/views/sale_order.xml]:
```suggestion
                <field name="inverse_currency_rate" attrs="{'invisible': [('sale_show_currency_rate', 'in', ['no', 'normal'])]}" />
```

File: sale_order_currency_rate/views/sale_order.xml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="view_order_form" model="ir.ui.view">
+        <field name="name">sale.order.form</field>
+        <field name="model">sale.order</field>
+        <field name="inherit_id" ref="sale.view_order_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='currency_id']" position="after">
+                <field name="currency_rate" />
```
Comment (alan196) [inline sale_order_currency_rate/views/sale_order.xml]:
```suggestion
                <field name="currency_rate" attrs="{'invisible': [('sale_show_currency_rate', 'in', ['no', 'inverse'])]}" />
```

File: sale_order_currency_rate/models/sale_order.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        store=True,
+        digits=(12, 6),
+    )
+
+    sale_show_currency_rate = fields.Selection(
+        [("show", "Show"), ("hide", "Hide")],
```
Comment (alan196) [inline sale_order_currency_rate/models/sale_order.py]:
```suggestion
```

This is a related field, it doesn't need to add the string and selection information

File: sale_order_currency_rate/models/sale_order.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        readonly=True,
+        store=True,
+    )
+
+    inverse_currency_rate = fields.Float(
+        compute="_compute_currency_rate",
+        string="Inverse Currency",
+        readonly=True,
+        store=True,
+        digits=(12, 6),
```
Comment (alan196) [inline sale_order_currency_rate/models/sale_order.py]:
```suggestion
        digits="Currency Rate Precision"
```

And create a digital.precision record with data

**Review by alan196 — CHANGES_REQUESTED** [review]

File: sale_order_currency_rate/views/res_company.xml [hunk 10]
```diff
@@ -0,0 +1,12 @@
+<odoo>
+    <record id="view_res_company_form_inherit" model="ir.ui.view">
+        <field name="name">res.company.form.inherit</field>
+        <field name="model">res.company</field>
+        <field name="inherit_id" ref="base.view_company_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='company_registry']" position="after">
+                <field name="sale_show_currency_rate" />
```
Comment (alan196) [inline sale_order_currency_rate/views/res_company.xml]: You don't need to add the field in res.company view.

You need to add it to res.config.settings view in sales configuration.

File: sale_order_currency_rate/__manifest__.py [hunk 11]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2023 Jarsa
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl).
+
+{
+    "name": "Currency Rate",
```
Comment (alan196) [inline sale_order_currency_rate/__manifest__.py]:
```suggestion
    "name": "Currency Rate in Sale Order",
```

File: sale_order_currency_rate/models/sale_order.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+from odoo import api, fields, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    currency_rate = fields.Float(
+        compute="_compute_currency_rate",
+        string="Currency",
```
Comment (alan196) [inline sale_order_currency_rate/models/sale_order.py]:
```suggestion
```

The currency string may lead to confusion in end users, if you don't define the string parameter it will take the field name, in this case "Currency Rate" that is acceptable

File: sale_order_currency_rate/models/sale_order.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    )
+
+    inverse_currency_rate = fields.Float(
+        compute="_compute_currency_rate",
+        string="Inverse Currency",
```
Comment (alan196) [inline sale_order_currency_rate/models/sale_order.py]:
```suggestion
```

**Review by alan196 — CHANGES_REQUESTED** [review]

**Review by alan196 — APPROVED** [review]
Functional & technical review 👍

**Review by alan196 — APPROVED** [review]

**Review by rousseldenis — APPROVED** [review]
Code review

**Comment by rousseldenis** [comment]
@oscarjarsa Is this one ready ?

**Comment by oscarjarsa** [comment]
@rousseldenis this is ready to merge
